lib/cc/workspace.rb

```diff
@@ -27,8 +29,13 @@ def filter(exclude_paths)
 
       CLI.debug("workspace filter start", exclude_paths: exclude_paths)
 
-      @paths ||= Dir.glob("*", File::FNM_DOTMATCH)
-      @paths = @paths.flat_map { |name| expand(name, exclusions) }
+      exclusions.exclude_paths.each do |exclusion|
+        if exclusions.glob?(exclusion)
+          path_tree.exclude_paths(exclusions.expanded_glob(exclusion))
+        else
+          path_tree.exclude_paths([exclusion])
+        end
+      end
```

Comment:
Now that the complicated part has shifted from the list itself to the individual patterns, I don't think `Exclusions` is what we want. We probably want `[Exclusion]` here:

```rb
def filter(patterns)
  patterns.each do |pattern|
    exclusion = Exclusion.new(pattern)

    if exclusion.glob?
      path_tree.exclude_paths(exclusion.expand)
    else
      path_tree.exclude_paths([exclusion.raw]) # or something
    end
  end
end
```

Then my next suggestion would be to resolve the TDA violation by pushing `glob?` into `expand`

```rb
class Exclusion
  def expand
    return [pattern] unless glob?

    # expansion logic...
  end
end
```

Then my *next* suggesion is... do we really need both `Workspace` and `PathTree`? At that point, `Workspace` isn't doing too much, and the indirection may not be worth it.